src/rest_api/classes/person/widgets/lineage.clj

```diff
@@ -0,0 +1,349 @@
+                                    (flatten))]
+    (if (empty? queriedDataSupervisors) nil queriedDataSupervisors)))
+
+(defn- recurse-supervisors [queriedData]
+  (if (nil? queriedData) nil
+    (for [value queriedData
+          :let [supervisorsData (query-supervisors (:otherpersonid value)) 
+                data (flatten (conj supervisorsData (recurse-supervisors supervisorsData)))]] data)))
+
+(defn- query-supervisees [personid]
+  (let [db (d/db datomic-conn)
+        queriedDataSupervisees (->> (d/q '[:find ?personid ?personname ?otherpersonid ?otherpersonname (pull ?personsupervisedbyent [*])
+                                           :in $ ?personid
+                                           :where
+                                           [?person :person/id ?personid]
+                                           [?otherpersonent :person/supervised-by ?personsupervisedbyent]
+                                           [?person :person/standard-name ?personname]
+                                           [?personsupervisedbyent :person.supervised-by/person ?person]
+                                           [?otherpersonent :person/id ?otherpersonid]
+                                           [?otherpersonent :person/standard-name ?otherpersonname]] db personid)
+                                    (map (fn [oid]
+                                           (generate-map-nodes-edges oid)))
+                                    (flatten))]
+    (if (empty? queriedDataSupervisees) nil queriedDataSupervisees)))
+
+(defn- recurse-supervisees [queriedData]
+  (if (nil? queriedData)
+    nil
+    (for [value queriedData
+          :let [superviseesData (query-supervisees (:otherpersonid value)) 
+                data (flatten (conj superviseesData (recurse-supervisees superviseesData)))]] data)))
+
+(defn- query-this-person [person]
+  (let [db (d/entity-db person)
+        queriedDataThisPerson (->> (d/q '[:find ?personid ?personname 
+                                          :in $ ?person
+                                          :where
+                                          [?person :person/standard-name ?personname]
+                                          [?person :person/id ?personid]
+                                          ]
+                                        db (:db/id person))
+                                   (map (fn [oid]
+                                          (generate-map-this-person oid)))
+                                   (flatten))]
+    (if (empty? queriedDataThisPerson) nil queriedDataThisPerson)))
+
+(defn ancestors-data [person]
+  (let [db (d/entity-db person)
+        queriedDataThisPerson (query-this-person person)
+        thisPersonId (:personid (first queriedDataThisPerson))
+        thisPersonSupervisors (query-supervisors thisPersonId)
+        thisPersonSupervisees (query-supervisees thisPersonId)
+        edgesDirect
```

Comment:
if I understand correctly, the code below is building up a JSON structure?
If so, then better to build the data structure in clojure maps/vectors then pass the result to a json library fn, such as `cheshire.core/generate-string`

Reply:
That's true, and that makes sense, I didn't know that was an option when I wrote it.  If it's okay to keep it as it is, then it's working fine, if not I'll think over how to pass the structures around to generate the maps/vectors.  In the future I can write new code with generate-string.